tools/drivewatch.go

```diff
@@ -575,6 +579,122 @@ func (dw *driveWatch) needRingUpdate(rd ringData) bool {
 	return false
 }
 
+func (dw *driveWatch) checkMissingDispersionObjects(rd ringData, cancel chan struct{}) error {
+	policy := rd.p.Index
+	db, err := dw.getDbAndLock()
+	defer dw.dbl.Unlock()
+	if err != nil {
+		return err
+	}
+	tx, err := db.Begin()
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback()
+	rows, err := tx.Query(`
+	SELECT partition, account, container, object from dispersion_report_detail
+	WHERE dispersion_report_id = (
+		SELECT id FROM dispersion_report
+		WHERE policy = ? AND rtype = ? ORDER BY create_date DESC LIMIT 1)
+	`, policy, "object")
+	if err != nil {
+		dw.logger.Error("SELECT for max checkMissingDispersionObjects", zap.Error(err))
+		return err
+	}
+	defer rows.Close()
+	missingPartitions := map[int64]string{}
+	for rows.Next() {
+		var p int64
+		var a, c, o string
+		if err := rows.Scan(&p, &a, &c, &o); err == nil {
+			missingPartitions[p] = fmt.Sprintf("%s/%s/%s",
+				common.Urlencode(a), common.Urlencode(c), common.Urlencode(o))
+		} else {
+			dw.logger.Error("SELECT for checkMissingDispersion", zap.Error(err))
+		}
+	}
+	longMissingPartitions := []int64{}
+
+	for p := range missingPartitions {
+
+		rows, err := tx.Query(`
+			SELECT drd.partition
+			FROM dispersion_report dr LEFT JOIN dispersion_report_detail drd
+			ON dr.id = drd.dispersion_report_id AND drd.partition = ?
+			WHERE dr.policy = ? AND dr.rtype = ? AND dr.create_date > ?
+			`, p, policy, "object", time.Now().Add(-dw.rescueMissingPartAge))
+
+		if err != nil {
+			dw.logger.Error("SELECT for reports checkMissingDispersionObjects", zap.Error(err))
+			return err
+		}
+		defer rows.Close()
+		isReportWithPart := false
+		for rows.Next() {
+			var p sql.NullInt64
+			if err := rows.Scan(&p); err == nil {
+				if p.Int64 == 0 {
+					// if part not in report, all copies were found
+					isReportWithPart = true
+				}
+			} else {
+				dw.logger.Error("SELECT for reports checkMissingDispersionObjects scan", zap.Error(err))
+				continue
+			}
+		}
+		if !isReportWithPart {
+			longMissingPartitions = append(longMissingPartitions, p)
+		}
+	}
+	dw.logger.Info("Dispersion heal start", zap.Int("numPartitions", len(longMissingPartitions)))
+	for i := len(longMissingPartitions) - 1; i > 0; i-- { // shuffle
+		j := rand.Intn(i + 1)
+		longMissingPartitions[j], longMissingPartitions[i] = longMissingPartitions[i], longMissingPartitions[j]
+	}
+	for _, p := range longMissingPartitions {
+		select {
+		case <-cancel:
+			break
+		default:
+		}
+		nodes := rd.r.GetNodes(uint64(p))
```

Comment:
Shouldn't these vars have been uint64s all along? Or was there a reason for them to be int64s before?

Reply:
there isn't a sql.NullUInt64 . is that good enough?